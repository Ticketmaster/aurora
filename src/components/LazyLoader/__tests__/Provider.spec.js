import renderer from "react-test-renderer";

import {
  renderProviderComponent,
  createImgWithSrcset,
  createImgWithSrc,
  createDiv,
  PROPS as providerProps
} from "../__mocks__/utils.mocks";

describe("LazyLoaderProvider", () => {
  it("should load a low res, blurred version of the image passed prior to loading", () => {
    const component = renderProviderComponent({}, renderer.create, {
      createNodeMock: createImgWithSrc
    });

    expect(component).toMatchSnapshot();
  });

  it("should load a high res version of the image passed when the imageRef supports srcset", () => {
    const component = renderProviderComponent({}, renderer.create, {
      createNodeMock: createImgWithSrcset
    });

    const instance = component.getInstance();
    instance.load(true);

    expect(instance.state.imageRef).toMatchSnapshot();
  });

  it("should load a high res version of the image passed when the imageRef supports src", () => {
    const component = renderProviderComponent({}, renderer.create, {
      createNodeMock: createImgWithSrc
    });

    const instance = component.getInstance();
    instance.load(true);

    expect(instance.state.imageRef).toMatchSnapshot();
  });

  it("should load a high res version of the image passed when the imageRef supports a backgroundImage style", () => {
    const component = renderProviderComponent({}, renderer.create, {
      createNodeMock: createDiv
    });

    const instance = component.getInstance();
    const onLoadSpy = jest.spyOn(instance, "onLoad");
    instance.load(true);

    expect(onLoadSpy).toHaveBeenCalled();
    instance.state.imageRef.current.onload.mockClear();
    expect(instance.state.imageRef).toMatchSnapshot();
  });

  it("should not load a high res version of the image passed when the load function is invoked with a value of false", () => {
    const component = renderProviderComponent({}, renderer.create, {
      createNodeMock: createImgWithSrcset
    });

    const instance = component.getInstance();
    const onLoadSpy = jest.spyOn(instance, "onLoad");
    instance.load(false);

    expect(onLoadSpy).not.toHaveBeenCalled();
    expect(component).toMatchSnapshot();
  });

  it("should not error when calling the onLoad method if no ref is available", () => {
    const component = renderProviderComponent({}, renderer.create);

    const instance = component.getInstance();
    expect(() => instance.onLoad()).not.toThrow();
  });

  it("should invoke componentDidUpdate and not setState when src equals prevSrc", () => {
    const component = renderProviderComponent({}, renderer.create, {
      createNodeMock: createImgWithSrcset
    });

    const instance = component.getInstance();
    const setStateSpy = jest.spyOn(instance, "setState");
    instance.componentDidUpdate(providerProps);

    expect(setStateSpy).not.toHaveBeenCalled();
    expect(instance.state).toMatchSnapshot();
  });

  it("should invoke componentDidUpdate and set state to a low res, blurred version of the image when the src changes", () => {
    const component = renderProviderComponent({}, renderer.create, {
      createNodeMock: createImgWithSrcset
    });

    const instance = component.getInstance();
    const setStateSpy = jest.spyOn(instance, "setState");
    instance.load(true);

    component.update(
      renderProviderComponent({ src: "https://placepuppy.com/g/" }, x => x, {
        createNodeMock: createImgWithSrcset
      })
    );

    expect(setStateSpy).toHaveBeenCalled();
    expect(component.toJSON()).toMatchSnapshot();
  });
});
