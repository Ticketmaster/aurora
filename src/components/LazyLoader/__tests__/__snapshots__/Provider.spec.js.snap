// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LazyLoaderProvider should invoke componentDidUpdate and not setState when src equals prevSrc 1`] = `
Object {
  "backgroundRef": null,
  "imageRef": Object {
    "current": Object {
      "src": "",
      "srcset": "",
      "style": Object {},
    },
  },
  "isLoaded": false,
  "load": [Function],
  "src": "https://placekitten.com/g/40/40",
  "style": Object {
    "filter": "blur(10px)",
    "transition": "80ms filter linear",
  },
}
`;

exports[`LazyLoaderProvider should invoke componentDidUpdate and set state to a low res, blurred version of the image when the src changes 1`] = `
<img
  alt="Test Kitten"
  height={400}
  src="https://placepuppy.com/g//40/40"
  style={
    Object {
      "filter": "blur(10px)",
      "transition": "80ms filter linear",
    }
  }
  width={400}
/>
`;

exports[`LazyLoaderProvider should load a high res version of the image passed when the imageRef supports src 1`] = `
Object {
  "current": Object {
    "onload": [Function],
    "src": "https://placekitten.com/g/400/400",
    "style": Object {},
  },
}
`;

exports[`LazyLoaderProvider should load a high res version of the image passed when the imageRef supports srcset 1`] = `
Object {
  "current": Object {
    "onload": [Function],
    "src": "",
    "srcset": "https://placekitten.com/g/400/400 1x, https://placekitten.com/g/800/800 2x, https://placekitten.com/g/1200/1200 3x, https://placekitten.com/g/1600/1600 4x, https://placekitten.com/g/2000/2000 5x",
    "style": Object {},
  },
}
`;

exports[`LazyLoaderProvider should load a high res version of the image passed when the tag equals div, imageRef supports src or srcset, and backgroundRef supports a backgroundImage style 1`] = `
Object {
  "backgroundRef": Object {
    "current": Object {
      "style": Object {
        "backgroundImage": "url(https://placekitten.com/g/400/400)",
        "filter": "none",
      },
    },
  },
  "imageRef": Object {
    "current": Object {
      "onload": [Function],
      "src": "https://placekitten.com/g/400/400",
      "srcset": "",
      "style": Object {
        "filter": "none",
      },
    },
  },
}
`;

exports[`LazyLoaderProvider should load a low res, blurred version of the image passed prior to loading 1`] = `
<img
  alt="Test Kitten"
  height={400}
  src="https://placekitten.com/g/40/40"
  style={
    Object {
      "filter": "blur(10px)",
      "transition": "80ms filter linear",
    }
  }
  width={400}
/>
`;

exports[`LazyLoaderProvider should low res, blurred version of the image passed prior to loading when the tag equals div, imageRef supports src or srcset, and backgroundRef supports a backgroundImage style 1`] = `
<div
  isLoaded={false}
  load={[Function]}
  src="https://placekitten.com/g/40/40"
  style={
    Object {
      "filter": "blur(10px)",
      "transition": "80ms filter linear",
    }
  }
>
  <img
    alt="Test Kitten"
    height={400}
    src="https://placekitten.com/g/40/40"
    style={
      Object {
        "filter": "blur(10px)",
        "transition": "80ms filter linear",
      }
    }
    width={400}
  />
</div>
`;

exports[`LazyLoaderProvider should not load a high res version of the image passed when the load function is invoked with a value of false 1`] = `
<img
  alt="Test Kitten"
  height={400}
  src="https://placekitten.com/g/40/40"
  style={
    Object {
      "filter": "blur(10px)",
      "transition": "80ms filter linear",
    }
  }
  width={400}
/>
`;
